torchrl/modules/tensordict_module/world_models.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from math import sqrt
+
+import torch
+import torch.distributions as d
+import torch.nn as nn
+import torch.nn.functional as F
+
+from torchrl.modules.tensordict_module import TensorDictModule, TensorDictSequence
+
+__all__ = ["WorldModelWrapper"]
+
+
+class WorldModelWrapper(TensorDictSequence):
```

Comment:
Does this ever happen that we have a common backbone between the two? Should we consider this option?

Reply:
What do you mean? The reward modeler will mostly be used on top of the world Model.  The goal of WorldModelWrapper is to concatenate world model and reward model but keeping an easy access to both of them.

Reply:
In ActorCritic and things like that the `get_policy_operator` and similar functions isolate the components that you need to run it on a tensordict collected from the replay buffer.
If WorldModelWrapper is a sequence where the reward model depends upon the `world_modeler_operator` then the fact that `get_reward_operator` returns just the reward model is counter intuitive.
I'm not super in favour of methods like
```python
def get_module(self):
    return self.module
```
Since you did not write `self._module` but `self.module`, module is considered public. Hence there is no advantage in having a `get_module` function.

We could think about the following settings:
1. reward model is based on world model: then `get_reward_model` should return what is needed to compute the reward (this should be documented in the docstring if it's the general setting)
```python
def get_reward_model(self):
    if isinstance(self.wold_model_operator, TensorDictSequence):
        sub_world_model = self.wold_model_operator.select_subsequence(out_keys=self.reward_model.in_keys)
   else:
        sub_world_model = self.wold_model_operator
    return TensorDictSequence(sub_world_model, selef.reward_model)
```
2. (*see below) reward model is a callable: this will require us to adapt TensorDictModule slightly, by automatically doing something like
```python
class callable_module(nn.Module):
    def __init__(self, callable):
        super().__init__()
        self.callable = callable

    def forward(self, *args):
        return self.callable(*args)


class TensorDictModule(...):
    def __init___(self, module, ...):
        if not isinstance(module, nn.Module):
           module = callable_module(module)
```
3. reward and world model are independent: then we don't need the get_funcitons IMO
4. there is a backbone
```python
class WorldModel(...):
    def __init__(self, backbone, obsmodel, rewardmodel):
        ...

    def get_obs_model(self):
        return TensorDictSequence(self.backbone, self.obsmodel)

    def get_reward_model(self):
        return TensorDictSequence(self.backbone, self.rewardmodel)

```

* Regarding (2.) let's consider cases where the reward model is actually known: in some instances the reward function is a differentiable function given a priori.

Reply:
For me, the idea in being able to retrieve the reward model is that we might want to compute the reward on some precomputed states. The thing with having this standalone reward model is that we loose this functionality.

Reply:
Wouldn't it be possible to compute directly the reward from the obervation and the action. The goal of this module is to easily be able to separate the state generator and the reward predictor. We could of course use a TDSequence and then do tdseq[0] and tedseq[1] but this fills less intuitive for the user. The thing is i dont see why this would be less relevant than something like ActorCriticWrapper?

Reply:
I agree that ActorCriticWrapper does the same thing :) But for context it has corresponding classes that do not do the same thing.
We can keep this, I'm happy with it for now. 
We will always be able to remove those classes if we feel they bring little advantage (usually we should try to avoid classes that just rename another class in a specific context like those two).